File: src/main/java/com/mercateo/jsonschema/generictype/GenericParameterizedType.java
package com.mercateo.jsonschema.generictype;

import static java.util.Objects.requireNonNull;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import java.util.Objects;

import com.googlecode.gentyref.GenericTypeReflector;

public final class GenericParameterizedType<T> extends GenericTypeAbstract<T, ParameterizedType> {

    GenericParameterizedType(ParameterizedType type, Class<T> rawType) {
        super(rawType, type);
    }

    @Override
    public String getSimpleName() {
        return rawType.getSimpleName();
    }

    @Override
    public GenericType<?> getContainedType() {
        Type[] actualTypeArguments = type.getActualTypeArguments();
        if (actualTypeArguments.length > 1) {
            throw new IllegalStateException(type + " not supported for subtyping");
        }
        return GenericType.of(actualTypeArguments[0], rawType);
    }

    @Override
    public GenericType<? super T> getSuperType() {
        final Class<? super T> superclass = rawType.getSuperclass();
        return superclass != null ? GenericType.of(GenericTypeReflector.getExactSuperType(type,
                superclass), superclass) : null;
    }

    @Override
    public String toString() {
        return type.toString();
    }

}

File: src/main/java/com/mercateo/jsonschema/generictype/GenericParameterizedType.kt
package com.mercateo.jsonschema.generictype

import com.googlecode.gentyref.GenericTypeReflector

import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

internal class GenericParameterizedType<T>(type: ParameterizedType, rawType: Class<T>) : GenericTypeAbstract<T, ParameterizedType>(rawType, type) {

    override val simpleName: String
        get() = type.typeName

    override val name: String
        get() = type.typeName

    override val containedType: GenericType<*>
        get() {
            val actualTypeArguments = type.actualTypeArguments
            if (actualTypeArguments.size > 1) {
                throw IllegalStateException(type.toString() + " not supported for subtyping")
            }
            return GenericType.of(actualTypeArguments[0], rawType)
        }

    override fun toString(): String {
        return type.toString()
    }

}
